### DISCUSSION TITLE
File Upload Handling

### DISCUSSION THREAD

**Discussion by iamganeshagrawal** [wailsapp/wails#773] [Q&A]
I have to create a CRM portal and I want to handle file upload via html `input[type=file]` so it is possible to pass a file to wails backend binded function. 

If yes then how?

**Comment by leaanthony** [comment]
I don't personally know but I'm sure someone has tried this. What have you tried so far?

**Reply by iamganeshagrawal** [reply]
Currently I am converting file to base64 dataurl in frontend and sending it as string to backend and in backend converting back using `https://github.com/vincent-petithory/dataurl` but this method taking a long time.

**Reply by leaanthony** [reply]
there might be something that could be done with service workers and byte arrays? Your best bet is just to send the filename to the backend and get Go to load it. If the file input doesn't give you the full path, potentially use the system dialog to get it.

**Reply by ghost** [reply]
I got full path using system dialog ( https://wails.io/docs/reference/runtime/dialog )

**Reply by illud** [reply]
@utadachiren do you have and example please?

**Reply by SheetJSDev** [reply]
[Complete example](https://docs.sheetjs.com/docs/demos/desktop#wails) of reading and writing file data.  @leaanthony please let us know if the example can be improved.  The Wails docs do not mention typed arrays or `Uint8Array` or `ArrayBuffer` at all.

On macOS Monterey with Wails `v2.0.0-beta.44.2` in a Svelte+TS app generated with `wails init -t svelte-ts`, the `input[type=file]` elements have no effect.  The File Input element appears but the button does not perform any action.  So the demo creates an independent button with event handler.

Assuming you have the raw data as a `Uint8Array` in JS and want a `[]byte` in Go, there are currently two ways of passing data:

a) Base64 string: encode in JS, pass string, Go receives string and decodes

b) Array of numbers: convert the `Uint8Array` with `Array.from(data)`, pass array, Go receives `[]byte`

**Reply by leaanthony** [reply]
Very happy to work with you on this @SheetJSDev! 
- The bug with `input[type=file]` on Mac is a [known issue](https://github.com/wailsapp/wails/issues/1331) (you have to implement it yourself in Obj-C!).
- The example on the site, I'd not use the `window['go']['main']['App']['ReadFile']();` method of calling methods - the best way to do this (as this way will break in the future) is to import the [generated wrappers](https://wails.io/docs/next/howdoesitwork#calling-bound-go-methods).

I would love your feedback on how the file handling could be better and the handling of `ArrayBuffer`. It's not something I'm overly familiar with and would love to improve it 🙏

**Reply by SheetJSDev** [reply]
Will raise separate issues on the matter

**Reply by moeyashi** [reply]
My example

go

```go
func (a *App) SelectFile() string {
	file, err := runtime.OpenFileDialog(a.ctx, runtime.OpenDialogOptions{})
	if err != nil {
		return err.Error()
	}
	return file
}
```

js

```js
const fileElement = document.getElementById("your file input id");
fileElement .onclick = () => {
  try {
    window.go.main.App.SelectFile()
      .then((result) => {
        fileElement .value = result;
      })
      .catch((err) => {
        console.error(err);
      });
  } catch (err) {
    console.error(err);
  }
}
```

**Reply by KiddoV** [reply]
@moeyashi I don't think your method in JS will work since for security reason the value of input with ``type="file"`` can only be set by the user, not programmatically.
This line => ``fileElement.value = result;`` will throws:
```error
DOMException: Failed to set the 'value' property on 'HTMLInputElement': This input element accepts a filename, which may only be programmatically set to the empty string.
```

**Reply by SheetJSDev** [reply]
There is a programmatic way of doing this with `DataTransfer`, but unfortunately it does not work in WebKit / Safari:

```js
fileElement.onclick = async (evt) => {
  evt.preventDefault();
  try {
    // assuming go SelectFile returns a path string
    const path = await window.go.main.App.SelectFile();
    var dt = new DataTransfer();
    dt.items.add(new File(["abc"], path));
    fileElement.files = dt.files;
  }
}
```

Demo https://jsfiddle.net/2wuan7xy/

In Chromium, the "No file chosen" string is updated to show the path string

**Reply by cyberxnomad** [reply]
> My example
> 
> go
> 
> ```go
> func (a *App) SelectFile() string {
> 	file, err := runtime.OpenFileDialog(a.ctx, runtime.OpenDialogOptions{})
> 	if err != nil {
> 		return err.Error()
> 	}
> 	return file
> }
> ```
> 
> js
> 
> ```js
> const fileElement = document.getElementById("your file input id");
> fileElement .onclick = () => {
>   try {
>     window.go.main.App.SelectFile()
>       .then((result) => {
>         fileElement .value = result;
>       })
>       .catch((err) => {
>         console.error(err);
>       });
>   } catch (err) {
>     console.error(err);
>   }
> }
> ```

great. resolve my problem. <3

**Comment by slo-loris** [comment]
The uploaded file has to go into a conversion process where it has to be converted to a string at JS end and has to be unmarshalled to a []byte at GO end to read the file. The conversion to string is required due to a strange behavior that I noticed while passing an array from JS to Go. All arrays passed from JS are being read as a map in GO. From my understanding this maybe because in JS, an array is basically a subset of the JS object and the counterpart of JS object in GO is a map. My workaround:

JS: (file -> array buffer -> unit8array -> string)

```typescript
const file = e.target.files[0]
Promise.resolve(file.arrayBuffer())).then((ab: ArrayBuffer) => {
  const u = new Uint8Array(ab);
  SendFileFromJStoGo(`[${u.toString()}]`);
});
```

GO: (string -> []byte -> file)

```go
SendFileFromJStoGo(blob string) {
  file := []byte{}
  if err := json.Unmarshal([]byte(file), &zipFile); err != nil {
    log.Fatal(err)
  }
}
```

For multiple file uploads:

JS: (file -> array buffer -> unit8array -> string)

```typescript
const files = Array.from(e.target.files)
Promise.all(files.map((file) => file.file.arrayBuffer())).then((ab: ArrayBuffer[]) => {
  const f = files.map((file) => {
    const u = new Uint8Array(file);
    return `[${u.toString()}]`;
  });
  SendFileFromJStoGo(f);
});
```

GO: (string -> []byte -> file)

```go
SendFileFromJStoGo(blobs []string) {
  for _, file := range blobs {
    file := []byte{}
    if err := json.Unmarshal([]byte(file), &zipFile); err != nil {
      log.Fatal(err)
    }

  }
}
```

You may notice that for conversion of unit8array into string in JS, I have used template literals & `toString()` method instead of JSON.stringify. This is again because of the strange behaviour of JS stringifying an unit8Array into a JS object instead of an array.

**Comment by mholt** [comment]
I'm also trying to figure out how to do this -- it's not an urgent blocking matter on my application, but it would be nice to support file drag-n-drop.

I guess I'm not clear on how wails would turn a multipart HTTP form into a native function call. It'd be pretty cool if we could make a struct with a `io.Reader` field and have that read from the file input.

(I admit I haven't tried this yet, but I assume it's not currently implemented, based on this discussion.)

**Comment by leaanthony** [comment]
The Optimus project has DnD. There's nteresting discussions on this topic here: https://github.com/microsoft/vscode/issues/115807
https://stackoverflow.com/questions/27034897/is-there-a-way-to-pass-an-arraybuffer-from-javascript-to-java-on-android

**Comment by guiklimek** [comment]
#### Context
I needed to upload TLS files in a Wails + TS React software.

#### Solution
I did the trick inspired by the [@slo-loris solution](https://github.com/wailsapp/wails/discussions/773#discussioncomment-4028611):
```typescript
const file: File = event.target.files[0];
file.arrayBuffer().then((fileBuffer: ArrayBuffer) => {
    const decoder: TextDecoder = new TextDecoder("utf-8");
    return decoder.decode(fileBuffer);
}).then((filecontent: string) => {
    // Do what you want with the decoded file in a string
});
```
Source: https://github.com/Ambre-io/sparkplugui/blob/main/frontend/src/components/containers/FormCard.tsx#L44.

#### Explanations

- After the `ArrayBuffer` step, I'm not using a `Uint8Array` because it makes the encoding layers complicated and I can't get back on my feet on the backend side.

- Instead, I use the [TextDecoder](https://developer.mozilla.org/en-US/docs/Web/API/TextDecoder) because I read positive comments and it seems to work well.

- The `decode` function gives me the file in a `string` which is convenient to send and use by the backend.

#### Best Practices & Security
1. I'm not sure it's a good thing to put the files in plaintext in memory, although the file is on the host machine in any case.

2. I'm not sure that this is a well-optimized solution, I imagine that a byte array would do better.
